### PR TITLE
Check if object to be deleted actually exists

### LINKED ISSUE #3463: Trying to delete no songs/maps, strange behaviours happen
## Issue
Trying to delete "nothing" with empty song list or with empty search song results makes strange behaviours with the entire song list _(is it called carousel? just to know)_, **and reopening the game everything gets fixed**.

## How to reproduce it
(Just like I said a few lines above)

## Screenshots
Pressing "Yes, totally. Delete it" this happens:
screenshot 1

Getting back to the song list, this is how its shown:
screenshot 2

Got back to the menu and reopening the song list, basically, remains stuck on this "empty window"
screenshot 3

And trying the _Home_ button just makes an error, and after pressing it like 3-4 times, the game crashes:
screenshot 4

The error is in the Italian language, but I put it anyway: everything can be useful in these cases.

## Logs
_Logs will be added soon: I have to get them back. In the meantime, you can try to reproduce the issue._

## Any Solutions?
Maybe trying to disable the "Option" button when there are no maps shown.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/63655239-trying-to-delete-no-songs-maps-strange-behaviours-happen?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by plankp**
Fix #3463 

The bug happens because the query `First` will expects there to be something, but there are no beatmaps in the user's carousel.

### CODE REVIEW

**Comment by peppy** [comment]
Not sure if this is the correct fix. It's worth figuring out how this got called with a non-existent item in the first place, and see if that is the better place to resolve the issue.

**Comment by plankp** [comment]
Would it be better if the `Delete Beatmap` button just slides away and does nothing?

**Comment by plankp** [comment]
The latest commit will cause the *Confirm deletion* overlay to not show up.

Keeping the `FirstOrDefault` code just in case some code does call that directly (as a safety measure?)

File: osu.Game/Screens/Select/SongSelect.cs [hunk 1]
```diff
@@ -537,6 +537,8 @@ private void carouselBeatmapsLoaded()
         private void delete(BeatmapSetInfo beatmap)
         {
             if (beatmap == null) return;
```
[the remark this answers is not in the document]
Comment (plankp) [reply osu.Game/Screens/Select/SongSelect.cs]: Why does testing `ID <= 0` work? (I am not familiar with the `BeatmapSetInfo` class)

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Select/SongSelect.cs]: The ID comes from the database. If the set doesn't have one it isn't stored to the database.

[the remark this answers is not in the document]
Comment (plankp) [reply osu.Game/Screens/Select/SongSelect.cs]: Ok, I will change it to that

File: osu.Game/Database/ArchiveModelManager.cs [hunk 2]
```diff
@@ -286,9 +286,10 @@ public void Delete(TModel item)
             using (ContextFactory.GetForWrite())
             {
                 // re-fetch the model on the import context.
-                var foundModel = queryModel().Include(s => s.Files).ThenInclude(f => f.FileInfo).First(s => s.ID == item.ID);
+                var foundModel = queryModel().Include(s => s.Files).ThenInclude(f => f.FileInfo).FirstOrDefault(s => s.ID == item.ID);
 
-                if (foundModel.DeletePending) return;
+                // Test for null since FirstOrDefault will return null if nothing is found
+                if (foundModel == null || foundModel.DeletePending) return;
```
Comment (peppy) [inline osu.Game/Database/ArchiveModelManager.cs]: While we're here, let's make this Delete function return `false` on early exit (when no operation was performed).

Comment (plankp) [reply osu.Game/Database/ArchiveModelManager.cs]: ok

**Comment by plankp** [comment]
Well... the test somehow resolved itself.

Should be ready

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**